torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -353,119 +353,137 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
   TORCH_INTERNAL_ASSERT(
-      tpMessage.payloads.size() == 1,
-      "message expected to contain 1 payload, whereas it contained ",
+      tpMessage.payloads.size() == 4,
+      "message expected to contain 4 payloads, whereas it contained ",
       tpMessage.payloads.size(),
       " payloads");
-  std::vector<char> payload(tpMessage.payloads[0].length);
-  tpMessage.payloads[0].data = (uint8_t*)(payload.data());
+
   TORCH_INTERNAL_ASSERT(
-      tpMessage.metadata.size() == 2 * sizeof(int64_t),
-      "message metadata must be ",
-      2 * sizeof(int64_t),
-      " bytes, whereas it is ",
-      tpMessage.metadata.size(),
+      tpMessage.payloads[0].length == sizeof(MessageType),
+      "first payload expected to contain ",
+      sizeof(MessageType),
+      " bytes, whereas it contained ",
+      tpMessage.payloads[0].length,
       " bytes");
-  int64_t mtypei, mId;
-  memcpy(&mtypei, tpMessage.metadata.data(), sizeof(int64_t));
-  memcpy(&mId, tpMessage.metadata.data() + sizeof(int64_t), sizeof(int64_t));
-  MessageType mType = static_cast<MessageType>(mtypei);
+  buffers.type = std::make_unique<MessageType>();
+  tpMessage.payloads[0].data = buffers.type.get();
 
-  // Tensors
-  std::vector<torch::Tensor> tensors;
-  tensors.reserve(tpMessage.tensors.size());
-  for (tensorpipe::Message::Tensor& tpTensor : tpMessage.tensors) {
-    const std::string& metadata = tpTensor.metadata;
-    size_t metadataPos = 0;
-    auto metaDataReadFunc = [&](char* buf, size_t n) -> size_t {
-      if (metadataPos >= metadata.size() || n == 0) {
-        return 0;
-      }
-      size_t toCopy = std::min(n, metadata.size() - metadataPos);
-      memcpy(buf, metadata.data() + metadataPos, toCopy);
-      metadataPos += toCopy;
-      return toCopy;
-    };
+  TORCH_INTERNAL_ASSERT(
+      tpMessage.payloads[1].length == sizeof(int64_t),
+      "second payload expected to contain ",
+      sizeof(int64_t),
+      " bytes, whereas it contained ",
+      tpMessage.payloads[1].length,
+      " bytes");
+  buffers.id = std::make_unique<int64_t>();
+  tpMessage.payloads[1].data = buffers.id.get();
 
-    auto sectionReadFunc = [&](const std::string& ename) -> at::DataPtr {
-      TORCH_INTERNAL_ASSERT(ename == "0", "single tensor ename must be \"0\"");
-      // TODO: CUDA memory allocation
-      return at::getCPUAllocator()->allocate(tpTensor.length);
-    };
+  // FIXME The two resizes below zero out the vectors, which is not needed.
 
-    torch::jit::Unpickler unpickler(
-        metaDataReadFunc, nullptr, nullptr, sectionReadFunc, {});
-    c10::IValue ival = unpickler.parse_ivalue();
-    at::Tensor rpcTensor = ival.toTensor();
-    tpTensor.data = (uint8_t*)(rpcTensor.data_ptr());
-    tensors.emplace_back(std::move(rpcTensor));
+  buffers.payload.resize(tpMessage.payloads[2].length);
+  tpMessage.payloads[2].data = buffers.payload.data();
+
+  buffers.pickle.resize(tpMessage.payloads[3].length);
+  tpMessage.payloads[3].data = buffers.pickle.data();
```

Comment:
Shall we have some `constexpr` for these indices instead of hard coding them?

Reply:
Good call!